src/jit/lower.cpp

```diff
@@ -2390,12 +2390,31 @@ GenTree* Lowering::LowerDelegateInvoke(GenTreeCall* call)
 
 GenTree* Lowering::LowerIndirectNonvirtCall(GenTreeCall* call)
 {
-#ifdef _TARGET_X86_
+#if defined(_TARGET_X86_)
     if (call->gtCallCookie != nullptr)
     {
-        NYI_X86("Morphing indirect non-virtual call with non-standard args");
+        // x86 Passes the P/Invoke cookie on the stack after all other arguments (if any). Add
+        // a putarg_stk here and remove the cookie operand from the call.
+        GenTree* const cookie = call->gtCallCookie;
+        call->gtCallCookie    = nullptr;
+
+        assert(cookie->gtType == TYP_I_IMPL);
+
+        const unsigned slotNum  = call->fgArgInfo->GetNextSlotNum();
+        const unsigned numSlots = 1;
+
+        GenTree* putArg =
+            new (comp, GT_PUTARG_STK) GenTreePutArgStk(GT_PUTARG_STK, cookie->gtType, cookie,
+                                                       slotNum PUT_STRUCT_ARG_STK_ONLY_ARG(numSlots) DEBUGARG(call));
+
+        if (cookie->InReg())
+        {
+            putArg->SetInReg();
+        }
+
+        BlockRange().InsertAfter(cookie, putArg);
     }
```

Comment:
Is this safe? You're creating a PUTARG_STK that isn't pointed to by any call, and isn't in any call arg table. Remorphing wouldn't know about it (maybe nobody remorphs after this?), and anybody else that uses the arg table wouldn't see it. It seems like you really want to add it as a PUTARG_STK in call argument morphing.